pkg/istio-agent/sds-agent.go

```diff
@@ -264,6 +273,66 @@ func gatewaySdsExists() bool {
 	return !os.IsNotExist(err)
 }
 
+// explicit code to determine the root CA to be configured in bootstrap file.
+// It may be different from the CA for the cert server - which is based on CA_ADDR
+// Replaces logic in the template:
+//                 {{- if .provisioned_cert }}
+//                  "filename": "{{(printf "%s%s" .provisioned_cert "/root-cert.pem") }}"
+//                  {{- else if eq .pilot_cert_provider "kubernetes" }}
+//                  "filename": "./var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+//                  {{- else if eq .pilot_cert_provider "istiod" }}
+//                  "filename": "./var/run/secrets/istio/root-cert.pem"
+//                  {{- end }}
+//
+// In addition it deals with the case the XDS server is on port 443, expected with a proper cert.
+// /etc/ssl/certs/ca-certificates.crt
+//
+// TODO: additional checks for existence. Fail early, instead of obscure envoy errors.
+func (sa *Agent) FindRootCAForXDS() string {
+	if sa.cfg.XDSRootCerts != "" {
+		return sa.cfg.XDSRootCerts
+	} else if _, err := os.Stat("./etc/certs/root-cert.pem"); err == nil {
+		// Old style - mounted cert. This is used for XDS auth only,
+		// not connecting to CA_ADDR because this mode uses external
+		// agent (Secret refresh, etc)
+		return "./etc/certs/root-cert.pem"
+	} else if sa.secOpts.PilotCertProvider == "kubernetes" {
+		// Using K8S - this is likely incorrect, may work by accident.
+		// API is alpha.
+		return k8sCAPath
+	} else if sa.secOpts.ProvCert != "" {
+		// This was never completely correct - PROV_CERT are only intended for auth with CA_ADDR,
+		// and should not be involved in determining the root CA.
+		return sa.secOpts.ProvCert + "/root-cert.pem"
+	} else {
+		// PILOT_CERT_PROVIDER - default is istiod
+		// This is the default - a mounted config map on K8S
+		return path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)
+	}
+}
+
+// Find the root CA to use when connecting to the CA (Istiod or external).
+//
+func (sa *Agent) FindRootCAForCA() string {
+	if sa.cfg.CARootCerts != "" {
+		return sa.cfg.CARootCerts
+	} else if sa.secOpts.PilotCertProvider == "kubernetes" {
+		// Using K8S - this is likely incorrect, may work by accident.
+		// API is alpha.
+		return k8sCAPath // ./var/run/secrets/kubernetes.io/serviceaccount/ca.crt
+	} else if sa.secOpts.PilotCertProvider == "custom" {
+		return security.DefaultRootCertFilePath // ./etc/certs/root-cert.pem
```

Comment:
Why is the logic different order from above? If its a real reason, should have comments, since it looks unintended

Reply:
Sorry, blind today - what is different ? Just the env variable should be.

Reply:
the ordering is different, and there is not /etc/certs/root-cert entry?